kong/plugins/ldap-auth/access.lua

```diff
@@ -50,17 +50,28 @@ local function ldap_authenticate(given_username, given_password, conf)
 
   sock:settimeout(conf.timeout)
 
-  ok, err = sock:connect(conf.ldap_host, conf.ldap_port)
+  local opts = {}
+
+  -- keep TLS connections in a separate pool to avoid reusing non-secure
+  -- connections and vica versa, because StartTLS use the same port 
+  if conf.start_tls then
+    opts.pool = conf.ldap_host .. ":" .. conf.ldap_port .. ":starttls"
+  end
+
+  ok, err = sock:connect(conf.ldap_host, conf.ldap_port, opts)
   if not ok then
     kong.log.err("failed to connect to ", conf.ldap_host, ":",
                    tostring(conf.ldap_port), ": ", err)
     return nil, err
   end
 
   if conf.start_tls then
-    local success, err = ldap.start_tls(sock)
-    if not success then
-      return false, err
+    -- convert connection to a StarTLS connection only if it is a new connection
+    if conf.start_tls and sock:getreusedtimes() == 0 then
+      local success, err = ldap.start_tls(sock)
+      if not success then
+        return false, err
```

Comment:
On failure, and by convention, our Lua functions should return `nil, err`, and not `false`.

Reply:
Same here, didn't want to change existing code which was working but thanks for the notice!